python/cuml/sparse/linalg/blopex_lobpcg.py

```diff
@@ -0,0 +1,647 @@
+#
+# Copyright (c) 2019-2020, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""
+Locally Optimal Block Preconditioned Conjugate Gradient Method (LOBPCG).
+References
+----------
+.. [1] A. V. Knyazev (2001),
+       Toward the Optimal Preconditioned Eigensolver: Locally Optimal
+       Block Preconditioned Conjugate Gradient Method.
+       SIAM Journal on Scientific Computing 23, no. 2,
+       pp. 517-541. http://dx.doi.org/10.1137/S1064827500366124
+.. [2] A. V. Knyazev, I. Lashuk, M. E. Argentati, and E. Ovchinnikov (2007),
+       Block Locally Optimal Preconditioned Eigenvalue Xolvers (BLOPEX)
+       in hypre and PETSc.  https://arxiv.org/abs/0705.2626
+.. [3] A. V. Knyazev's C and MATLAB implementations:
+       https://github.com/lobpcg/blopex
+"""
+
+import cupy as cp
+from cupy.linalg import (solve, cholesky, inv, eigh)
+from cupyx.scipy.sparse import (spmatrix, issparse)
+# import warnings
+from warnings import warn
+
+__all__ = ['lobpcg']
+
+
+def _report_nonhermitian(M, name):
+    """
+    Report if `M` is not a hermitian matrix given its type.
+    """
+    from cupy.linalg import norm
+
+    md = M - M.T.conj()
+
+    nmd = norm(md, 1)
+    tol = 10 * cp.finfo(M.dtype).eps
+    tol = max(tol, tol * norm(M, 1))
+    if nmd > tol:
+        print('matrix %s of the type %s is not sufficiently Hermitian:' %
+              (name, M.dtype))
+        print('condition: %.e < %e' % (nmd, tol))
+
+
+def _as2d(ar):
+    """
+    If the input array is 2D return it, if it is 1D, append a dimension,
+    making it a column vector.
+    """
+    if ar.ndim == 2:
+        return ar
+    else:  # Assume 1!
+        aux = cp.array(ar, copy=False)
+        aux.shape = (ar.shape[0], 1)
+        return aux
+
+
```

Comment:
Are we able to include and use `_makeoperator` in here as well? It looks like this version follows the Scipy version pretty closely.

Reply:
Yes corey, will have to basically implement [this](https://github.com/scipy/scipy/blob/v1.5.2/scipy/sparse/linalg/interface.py) in cupy. (This will also take us a step closer in completing this [comparison table!](https://github.com/scipy/scipy/blob/v1.5.2/scipy/sparse/linalg/interface.py))

Reply:
at the time of writing it, sparsity was not as much of a priority as the feature itself; so was prioritizing the functionality for dense input. I think it's time I prioritize sparsity ✌️